feat: add `string/to-well-formed`

Pratik772846 opened the pull request:
Returns a well-formed string by replacing all lone surrogates

Fixes #1066

Resolves # 1066

## Description

> What is the purpose of this pull request?

This pull request:

-   Replaces all lone surrogates in a string and returns a well-formed string
-   If the string is already well formed then return a shallow copy of the string

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves # 
-   fixes #

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Pratik772846 committed:
feat: added @stdlib/string/to-well-formed

Returns a well formed string by replacing all lone surrogates

Fixes #1066

Pranavchiku commented on lib/node_modules/@stdlib/string/to-well-formed/README.md:
```diff
@@ -0,0 +1,108 @@
+
+# str2wellformed
+
+> Creates a new well-formed string.
```
```suggestion
> Create a new well-formed string.
```

Pranavchiku commented on lib/node_modules/@stdlib/string/to-well-formed/README.md:
```diff
@@ -0,0 +1,108 @@
+
+<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->
+
+<section class="related">
+
+
+</section>
```
We can put some related packages here.

Pranavchiku commented on lib/node_modules/@stdlib/string/to-well-formed/benchmark/benchmark.js:
```diff
@@ -0,0 +1,59 @@
+
+// MAIN //
+
+bench( pkg+'::str2wellformed', function benchmark( b ) {
```
```suggestion
bench( pkg, function benchmark( b ) {
```
We don't need this here, any specific reason for having this?

Pranavchiku commented on lib/node_modules/@stdlib/string/to-well-formed/docs/types/index.d.ts:
```diff
@@ -0,0 +1,44 @@
+// TypeScript Version: 4.1
+
+/**
+* Tests if a string is well-formed.
```
Incorrect description

Pranavchiku commented on lib/node_modules/@stdlib/string/to-well-formed/docs/types/test.ts:
```diff
@@ -0,0 +1,34 @@
+*/
+
+import str2wellformed = require( './index' );
+
+// TESTS //
```
```suggestion


// TESTS //
```

Pranavchiku replied:
Mek sure you have correct dev env setup

Pranavchiku commented on lib/node_modules/@stdlib/string/to-well-formed/examples/index.js:
```diff
@@ -0,0 +1,35 @@
+
+var str2wellformed = require( './../lib' );
+
+console.log( str2wellformed( '' ) );
+
```
```suggestion
console.log( str2wellformed( '' ) );
// => ''

```

I will prefer to do it for all console.log here

Pranavchiku commented on lib/node_modules/@stdlib/string/to-well-formed/lib/index.js:
```diff
@@ -0,0 +1,50 @@
+* str = "abc";
+* bool = str2wellformed( str );
+* // returns abc
+*
```
```suggestion
```

Pranavchiku commented on lib/node_modules/@stdlib/string/to-well-formed/lib/main.js:
```diff
@@ -0,0 +1,73 @@
+* // returns 'ab�'
+*/
+function str2wellformed( str ) {
+	var modifiedString = '';
+	var i;
```
```suggestion
	var modifiedString;
	var i;
	modifiedString = '';
```

Pranavchiku commented on lib/node_modules/@stdlib/string/to-well-formed/lib/main.js:
```diff
@@ -0,0 +1,73 @@
+		}
+		return modifiedString;
+	}
+	for ( i =0; i < str.length; i++ ) {
```
```suggestion
	for ( i = 0; i < str.length; i++ ) {
```

Pranavchiku commented on lib/node_modules/@stdlib/string/to-well-formed/lib/main.js:
```diff
@@ -0,0 +1,73 @@
+			modifiedString += '\uFFFD';
+		}
+		else {
+			modifiedString += str[i];
```
```suggestion
			modifiedString += str[ i ];
```

Pranavchiku commented on lib/node_modules/@stdlib/string/to-well-formed/package.json:
```diff
@@ -0,0 +1,72 @@
+      "primitive",
+      "object"
+    ]
+  }
```
incorrect indentation

Pranavchiku reviewed (changes requested):
Thanks @Pratik772846 for working on this. Requested changes.

Pratik772846 committed:
feat: add string/to-well-formed

requested changes done
Returns a well formed string by replacing all lone surrogates

Fixes #1066

Pratik772846 committed:
feat: add string/to-well-formed

requested changes done
Returns a well formed string by replacing all lone surrogates

Fixes #1066

Pratik772846 committed:
feat: add string/to-well-formed

requested changes done
Returns a well formed string by replacing all lone surrogates

Fixes #1066

Pranavchiku commented on lib/node_modules/@stdlib/string/to-well-formed/package.json:
```diff
@@ -0,0 +1,72 @@
+    "object"
+  ]
+}
+  
```
```suggestion
```

Pranavchiku committed:
chore: apply code review

Signed-off-by: Pranav <85227306+Pranavchiku@users.noreply.github.com>

Pranavchiku reviewed (approved):
LGTM.

Pranavchiku commented:
I am not so great in reviewing this kind of PRs, I let Athan or Phillip review this.

Pratik772846 commented:
@Planeshifter Can you please review these changes?

Planeshifter commented on lib/node_modules/@stdlib/string/to-well-formed/docs/repl.txt:
```diff
@@ -0,0 +1,31 @@
+    Parameters
+    ----------
+    str: string
+        String to test.
```
```suggestion
        Input string.
```

Planeshifter commented on lib/node_modules/@stdlib/string/to-well-formed/package.json:
```diff
@@ -0,0 +1,71 @@
+    "windows"
+  ],
+  "keywords": [
+    "stdlib",
+    "stdassert",
+    "assertion",
+    "assert",
+    "utilities",
+    "utility",
+    "utils",
+    "util",
+    "number",
+    "numeric",
+    "negative",
+    "is",
+    "isnumber",
+    "isnumeric",
+    "type",
+    "check",
+    "primitive",
+    "object"
```
```suggestion
    "stdlib",
    "string",
    "well-formed",
    "surrogate",
    "lone-surrogate",
    "utf-16",
    "unicode",
    "replacement",
    "character",
    "validation",
    "sanitization",
    "conversion",
    "transform",
    "utility",
    "utils"
```

Planeshifter commented on lib/node_modules/@stdlib/string/to-well-formed/test/test.js:
```diff
@@ -0,0 +1,46 @@
+});
+
+tape( 'the function returns a well-formed string by replacing lone surrogates with Unicode replacement character', function test( t ) {
+	t.equal(str2wellformed(''), '', 'returns an empty string for an empty input');
+	t.equal(str2wellformed('\uDBFF'), '�', 'replaces lone high surrogate with Unicode replacement character');
+	t.equal(str2wellformed('\uDBFFFF\uDBFF'), '�FF�', 'replaces multiple lone surrogates with Unicode replacement characters');
+	t.equal(str2wellformed('Hello \uDBFFWorld'), 'Hello �World', 'replaces lone high surrogate in a string with Unicode replacement character');
+	t.equal(str2wellformed('\uDBFF'), '�', 'replaces lone high surrogate with Unicode replacement character');
+	t.equal(str2wellformed('\uDBFFFFF'), '�FFF', 'replaces lone high surrogate with Unicode replacement character');
+	t.equal(str2wellformed('\uDC00\uDC00'), '��', 'replaces lone low surrogate with Unicode replacement character');
+	t.equal(str2wellformed('5\uDBFF'), '5�', 'replaces lone high surrogate with Unicode replacement character');
```
```suggestion
	t.equal( str2wellformed(''), '', 'returns an empty string for an empty input' );
	t.equal( str2wellformed('\uDBFF'), '�', 'replaces lone high surrogate with Unicode replacement character' );
	t.equal( str2wellformed('\uDBFFFF\uDBFF'), '�FF�', 'replaces multiple lone surrogates with Unicode replacement characters' );
	t.equal( str2wellformed('Hello \uDBFFWorld'), 'Hello �World', 'replaces lone high surrogate in a string with Unicode replacement character' );
	t.equal( str2wellformed('\uDBFF'), '�', 'replaces lone high surrogate with Unicode replacement character' );
	t.equal( str2wellformed('\uDBFFFFF'), '�FFF', 'replaces lone high surrogate with Unicode replacement character' );
	t.equal( str2wellformed('\uDC00\uDC00'), '��', 'replaces lone low surrogate with Unicode replacement character' );
	t.equal( str2wellformed('5\uDBFF'), '5�', 'replaces lone high surrogate with Unicode replacement character' );
```

Planeshifter commented on lib/node_modules/@stdlib/string/to-well-formed/package.json:
```diff
@@ -0,0 +1,71 @@
+{
+  "name": "@stdlib/string/to-well-formed",
+  "version": "0.0.0",
+  "description": "Convert a string to a well formed string",
```
```suggestion
  "description": "Convert a string to a well formed string.",
```

Planeshifter commented on lib/node_modules/@stdlib/string/to-well-formed/lib/index.js:
```diff
@@ -0,0 +1,49 @@
+'use strict';
+
+/**
+* Return a well formed string.
```
```suggestion
* Convert a string to a well formed string.
```

Planeshifter commented on lib/node_modules/@stdlib/string/to-well-formed/docs/types/index.d.ts:
```diff
@@ -0,0 +1,44 @@
+// TypeScript Version: 4.1
+
+/**
+* Returns a well formed string.
```
```suggestion
* Replaces all lone surrogates in order to create a new string which is well formed.
```

Planeshifter commented on lib/node_modules/@stdlib/string/to-well-formed/docs/types/index.d.ts:
```diff
@@ -0,0 +1,44 @@
+* Returns a well formed string.
+*
+* @param str - string to test
+* @returns returns a well formed string
```
```suggestion
* @returns new string which does not contain any lone surrogates
```

Planeshifter commented on lib/node_modules/@stdlib/string/to-well-formed/README.md:
```diff
@@ -0,0 +1,107 @@
+
+#### str2wellformed( str )
+
+Creates a new well-formed string.
```
```suggestion
Creates a new well-formed string by replacing all lone surrogates with the replacement character `\uFFFD`.
```

Planeshifter committed:
Apply suggestions from code review

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter commented:
@Pratik772846 Implementation looks fine; will merge after CI has passed. Thank you!
